File: emailService.js
const nodemailer = require('nodemailer');

/**
 * Validates required environment variables
 * @throws {Error} If required env vars are missing
 */
function validateConfig() {
  if (!process.env.SMTP_USER || !process.env.SMTP_PASS) {
    throw new Error('SMTP_USER and SMTP_PASS environment variables must be set');
  }
}

/**
 * Creates and configures the email transporter
 * @returns {nodemailer.Transporter} Configured transporter
 */
function createTransporter() {
  validateConfig();
  
  return nodemailer.createTransport({
    host: process.env.SMTP_HOST || 'smtp.gmail.com',
    port: parseInt(process.env.SMTP_PORT) || 587,
    secure: process.env.SMTP_SECURE === 'true',
    auth: {
      user: process.env.SMTP_USER,
      pass: process.env.SMTP_PASS
    }
  });
}

/**
 * Escapes HTML special characters to prevent XSS
 * @param {string} text - Text to escape
 * @returns {string} Escaped text
 */
function escapeHtml(text) {
  const map = {
    '&': '&amp;',
    '<': '&lt;',
    '>': '&gt;',
    '"': '&quot;',
    "'": '&#039;'
  };
  return String(text).replace(/[&<>"']/g, m => map[m]);
}

/**
 * Formats field name for display
 * @param {string} key - Field key
 * @returns {string} Formatted label
 */
function formatFieldLabel(key) {
  const labels = {
    nome: 'Nome',
    email: 'Email',
    telefono: 'Telefono',
    messaggio: 'Messaggio'
  };
  return labels[key] || key;
}

/**
 * Sends a confirmation email to the user
 * @param {string} toEmail - Recipient email address
 * @param {Object} formData - Form data submitted by the user
 * @returns {Promise<Object>} Email sending result
 */
async function sendConfirmationEmail(toEmail, formData) {
  const transporter = createTransporter();

  const mailOptions = {
    from: process.env.EMAIL_FROM || '"Form Submission" <noreply@example.com>',
    to: toEmail,
    subject: 'Conferma Ricezione Form - Grazie per la tua richiesta',
    html: `
      <html>
        <head>
          <style>
            body {
              font-family: Arial, sans-serif;
              line-height: 1.6;
              color: #333;
            }
            .container {
              max-width: 600px;
              margin: 0 auto;
              padding: 20px;
              background-color: #f9f9f9;
              border: 1px solid #ddd;
              border-radius: 5px;
            }
            .header {
              background-color: #4CAF50;
              color: white;
              padding: 10px;
              text-align: center;
              border-radius: 5px 5px 0 0;
            }
            .content {
              background-color: white;
              padding: 20px;
              margin-top: 10px;
              border-radius: 0 0 5px 5px;
            }
            .footer {
              text-align: center;
              margin-top: 20px;
              font-size: 12px;
              color: #888;
            }
            .data-label {
              font-weight: bold;
              color: #555;
            }
          </style>
        </head>
        <body>
          <div class="container">
            <div class="header">
              <h2>Conferma Ricezione</h2>
            </div>
            <div class="content">
              <p>Gentile utente,</p>
              <p>Abbiamo ricevuto correttamente il tuo form. Grazie per averci contattato!</p>
              <p>Ecco un riepilogo dei dati che ci hai inviato:</p>
              <ul>
                ${Object.entries(formData)
                  .map(([key, value]) => `<li><span class="data-label">${formatFieldLabel(key)}:</span> ${escapeHtml(value)}</li>`)
                  .join('')}
              </ul>
              <p>Ti risponderemo il prima possibile.</p>
              <p>Cordiali saluti,<br>Il Team</p>
            </div>
            <div class="footer">
              <p>Questa è una email automatica, si prega di non rispondere.</p>
            </div>
          </div>
        </body>
      </html>
    `,
    text: `
Gentile utente,

Abbiamo ricevuto correttamente il tuo form. Grazie per averci contattato!

Ecco un riepilogo dei dati che ci hai inviato:
${Object.entries(formData)
  .map(([key, value]) => `${key}: ${value}`)
  .join('\n')}

Ti risponderemo il prima possibile.

Cordiali saluti,
Il Team

---
Questa è una email automatica, si prega di non rispondere.
    `
  };

  try {
    const info = await transporter.sendMail(mailOptions);
    console.log('Email inviata con successo:', info.messageId);
    return {
      success: true,
      messageId: info.messageId
    };
  } catch (error) {
    console.error('Errore nell\'invio dell\'email:', error.message);
    throw new Error('Impossibile inviare l\'email');
  }
}

module.exports = {
  sendConfirmationEmail
};

File: email_service.py
"""
Email service module for sending automatic confirmation emails.
"""
import smtplib
from email.mime.text import MIMEText
from email.mime.multipart import MIMEMultipart
import os
import logging

logger = logging.getLogger(__name__)


def send_confirmation_email(recipient_email, recipient_name, message_content):
    """
    Send an automatic confirmation email to the user who submitted a message.
    
    Args:
        recipient_email (str): Email address of the recipient
        recipient_name (str): Name of the recipient
        message_content (str): The message content they submitted
        
    Raises:
        Exception: If email sending fails
    """
    # Email configuration from environment variables
    smtp_server = os.getenv('SMTP_SERVER', 'smtp.gmail.com')
    smtp_port = int(os.getenv('SMTP_PORT', '587'))
    sender_email = os.getenv('SENDER_EMAIL', 'noreply@example.com')
    sender_password = os.getenv('SENDER_PASSWORD', '')
    
    # In development/test mode, just log instead of actually sending
    if os.getenv('EMAIL_MODE', 'test') == 'test':
        logger.info(f"[TEST MODE] Email di conferma per {recipient_email}")
        logger.info(f"Destinatario: {recipient_name}")
        logger.info(f"Contenuto: {message_content}")
        return True
    
    # Create the email message
    msg = MIMEMultipart('alternative')
    msg['Subject'] = 'Conferma Ricezione Richiesta'
    msg['From'] = sender_email
    msg['To'] = recipient_email
    
    # Create email body (HTML version)
    html_body = f"""
    <html>
    <body style="font-family: Arial, sans-serif; padding: 20px;">
        <h2>Conferma Ricezione Richiesta</h2>
        <p>Gentile {recipient_name},</p>
        <p>Abbiamo ricevuto la tua richiesta e ti confermiamo che <strong>non è finita nel buco nero digitale</strong>! 🎯</p>
        <div style="background-color: #f5f5f5; padding: 15px; border-left: 4px solid #4CAF50; margin: 20px 0;">
            <h3>Dettagli della tua richiesta:</h3>
            <p><strong>Messaggio:</strong></p>
            <p style="white-space: pre-wrap;">{message_content}</p>
        </div>
        <p>Il nostro team esaminerà la tua richiesta e ti risponderà al più presto.</p>
        <p>Grazie per averci contattato!</p>
        <hr style="margin-top: 30px; border: none; border-top: 1px solid #ddd;">
        <p style="color: #666; font-size: 12px;">Questa è una conferma automatica. Si prega di non rispondere a questa email.</p>
    </body>
    </html>
    """
    
    # Create plain text version
    text_body = f"""
    Conferma Ricezione Richiesta
    
    Gentile {recipient_name},
    
    Abbiamo ricevuto la tua richiesta e ti confermiamo che non è finita nel buco nero digitale!
    
    Dettagli della tua richiesta:
    Messaggio: {message_content}
    
    Il nostro team esaminerà la tua richiesta e ti risponderà al più presto.
    
    Grazie per averci contattato!
    
    ---
    Questa è una conferma automatica. Si prega di non rispondere a questa email.
    """
    
    # Attach both versions
    part1 = MIMEText(text_body, 'plain')
    part2 = MIMEText(html_body, 'html')
    msg.attach(part1)
    msg.attach(part2)
    
    try:
        # Connect to SMTP server and send email
        with smtplib.SMTP(smtp_server, smtp_port) as server:
            server.starttls()
            if sender_password:
                server.login(sender_email, sender_password)
            server.send_message(msg)
        
        logger.info(f"Email di conferma inviata con successo a {recipient_email}")
        return True
        
    except Exception as e:
        logger.error(f"Errore nell'invio dell'email a {recipient_email}: {e}")
        raise
